File: VinitKumar/Problems/Milestone1/Problem11.cs
using System;
using VinitKumar.Utilities;

/*
 * Tasks - 
		1. Iterative solution only - [94. Binary Tree Inorder Traversal](https://leetcode.com/problems/binary-tree-inorder-traversal/description/)
		1. Iterative solution only - [144. Binary Tree Preorder Traversal](https://leetcode.com/problems/binary-tree-preorder-traversal/description/)
		1. Iterative solution only - [145. Binary Tree Postorder Traversal](https://leetcode.com/problems/binary-tree-postorder-traversal/description/)
 * */

namespace VinitKumar.Problems
{
	public class Problem11
	{
		public static void InOrderTraversal(TreeNode? root)
		{
			Stack<TreeNode> st = new();
			while(root != null || st.Count > 0)
			{
				while(root != null)
				{
					st.Push(root);
					root = root.left;
				}
				root = st.Pop();
				Console.WriteLine(root.val);
				root = root?.right;
			}
		}

        public static void PreOrderTraversal(TreeNode? root)
        {
			Stack<TreeNode> st = new();
			while(root != null || st.Count > 0)
			{
				while(root != null)
				{
                    Console.WriteLine(root.val);
                    st.Push(root);
					root = root.left;
				}
				root = st.Pop();
				root = root?.right;
			}

        }

        public static IList<int> PostorderTraversal(TreeNode? root)
        {
            Stack<TreeNode> st = new();
            IList<int> ans = [];
            if (root == null) return ans;
            TreeNode? lastVisited = null;

            while (root != null || st.Count > 0)
            {
                if (root != null)
                {
                    st.Push(root);
                    root = root.left;
                }
                else
                {
                    TreeNode? top = st.Peek();
                    if (top?.right != null && lastVisited != top?.right)
                    {
                        root = top?.right;
                    }
                    else
                    {
                        ans.Add((top?.val).GetValueOrDefault());
                        lastVisited = st.Pop();
                    }
                }
            }
            return ans;
        }
    }
}

File: VishnuThakral/Milestone1/Problem11.py
#Milestone 1 Problem11
#Combined Leetcode numbers: 94, 144, 145
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        stack = []
        current = root
        while current or stack:
            if current:
                stack.append(current)
                current = current.left
            else:
                p  = stack.pop()
                res.append(p.val)
                if p.right:
                    current = p.right
        return res

    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []

        stack = [root]
        res = []

        while stack:
            node = stack.pop()
            res.append(node.val)

            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)
        return res
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        def traverse(root):
            if root:
                if root.left:
                    traverse(root.left)
                if root.right:
                    traverse(root.right)
                res.append(root.val)
        res = []
        traverse(root)
        return res